packages/library/src/commonMain/kotlin/io/realm/RealmList.kt

```diff
@@ -16,166 +16,174 @@
-
-    override fun set(index: Int, element: E): E = facade.set(index, element)
-
-    override fun subList(fromIndex: Int, toIndex: Int): MutableList<E> =
-        facade.subList(fromIndex, toIndex)
-
-    override fun clear() = facade.clear()
 }
 
 /**
- * TODO
+ * Represents an unmanaged [RealmList] backed by a [MutableList] via class delegation.
  */
-private abstract class ListFacade<E> : MutableList<E>, AbstractMutableCollection<E>()
+private class UnmanagedListDelegate<E>(
+    list: MutableList<E> = mutableListOf()
+) : MutableList<E> by list
 
 /**
- * TODO
+ * Represents a managed [RealmList]. Its data will be persisted in Realm.
+ *
+ * [AbstractMutableList] provides enough default implementations on which we can rely. It can also
+ * be used as a delegate since it implements [MutableList].
  */
-private class UnmanagedListFacade<E> : ListFacade<E>() {
-
-    private val unmanagedList = mutableListOf<E>()
-
-    override val size: Int
-        get() = unmanagedList.size
-
-    override fun get(index: Int): E = unmanagedList[index]
-
-    override fun indexOf(element: E): Int = unmanagedList.indexOf(element)
-
-    override fun iterator(): MutableIterator<E> = unmanagedList.iterator()
-
-    override fun lastIndexOf(element: E): Int = unmanagedList.lastIndexOf(element)
-
-    override fun add(element: E): Boolean = unmanagedList.add(element)
-
-    override fun add(index: Int, element: E) = unmanagedList.add(index, element)
+private class ManagedListDelegate<E>(
+    private val listPtr: NativePointer,
+    metadata: RealmList.OperatorMetadata
+) : AbstractMutableList<E>() {
 
-    override fun addAll(index: Int, elements: Collection<E>): Boolean =
-        unmanagedList.addAll(index, elements)
-
-    override fun listIterator(): MutableListIterator<E> = unmanagedList.listIterator()
-
-    override fun listIterator(index: Int): MutableListIterator<E> =
-        unmanagedList.listIterator(index)
-
-    override fun removeAt(index: Int): E = unmanagedList.removeAt(index)
-
-    override fun set(index: Int, element: E): E = unmanagedList.set(index, element)
-
-    override fun subList(fromIndex: Int, toIndex: Int): MutableList<E> =
-        unmanagedList.subList(fromIndex, toIndex)
-}
-
-/**
- * TODO
- */
-private class ManagedListFacade<E>(
-    private val listPtr: NativePointer
-) : ListFacade<E>() {
+    private val operator = RealmList.Operator<E>(metadata)
 
     override val size: Int
-        get() = RealmInterop.realm_list_size(listPtr).toInt()
-
-    override fun get(index: Int): E = RealmInterop.realm_list_get(listPtr, index.toLong())
-
-    override fun indexOf(element: E): Int {
-        TODO("indexOf(element: E) - Not yet implemented")
+        get() {
+            operator.checkRealmClosed()
+            return RealmInterop.realm_list_size(listPtr).toInt()
+        }
+
+    override fun get(index: Int): E {
+        operator.checkRealmClosed()
+        rangeCheck(index)
```

Comment:
I think these are leftovers. I added the `rangeCheck`s to all methods using an `index` parameters only for testing purposes. It is only needed for the `addAll` with index method (due to the Kotlin Native bug). I guess it's OK to remove the checks for the others since Core does it for us.